archive/miu_language/basic.lean

```diff
@@ -0,0 +1,249 @@
+The system is defined by four rules of inference, one axiom, and an alphabet of three symbols:
+`M`, `I`, and `U`.
+
+Hofstadter's central question is: can the string "MU" be derived?
+
+It transpires that there is a simple decision procedure for this system. A string is derivable if
+and only if it starts with `"M"`, contains no other `"M"`s, and the number of `"I"`s in the string
+is congruent to 1 or 2 modulo 3.
+
+## The MIU System
+
+An _atom_ is any one of `M`, `I` or `U`. A _string_ is a finite sequence of zero or more symbols.
+To simplify notation, we write a sequence `[I,U,U,M]`, for example, as `IUUM`.
+
+The four rules of inference are:
+
+1. x`I` → x`IU`,
+2. `M`x → `M`xx,
+3. x`III`y → x`U`y,
+4. x`UU`y → xy,
+
+where the notation α → β is to be interpreted as 'if α is derivable, then β is derivable'.
+
+Additionally, we have an axiom
+
+* `MI` is derivable.
+
+In this file, the set of atoms and the set of derivable strings are represented as inductive types.
+-/
+
+
+namespace miu
+
+/-!
+### Basic data types
+-/
+
+/--
+Each MIU string consists of either an `M`, `I`, or `U`. Such an elementary unit is called an
+`miu_atom`. We represent `miu_atom` as an enumerated type.
+-/
+inductive miu_atom : Type
+| M : miu_atom
+| I : miu_atom
+| U : miu_atom
+
+open miu_atom 
+
+/--
+We show that the type `miu_atom` is inhabited, giving `M` (for no particular reason) as the default
+element.
+-/
+instance miu_atom_inhabited : inhabited miu_atom :=
+inhabited.mk M
+
+/--
+A simple function from `miu_atom` to `string`.
+-/
+def miu_atom.repr : miu_atom → string 
+| M := "M"
+| I := "I"
+| U := "U"
+
+/--
+A representation of an `miu_atom`.
+-/
+instance : has_repr miu_atom :=
+⟨λ u, u.repr⟩
+
+/--
+For simplicity, an `miustr` is just a list of `miu_atom`.
+-/
+def miustr := list miu_atom 
+
+/--
+We want to use list membership ...
+-/
+instance : has_mem miu_atom miustr :=
+  ⟨list.mem⟩
+
+/--
+... and list append.
+-/
+instance : has_append miustr :=
+⟨list.append⟩
+
+/--
+For display purposes, an `miustr` can be represented as a `string`.
+-/
+def miustr.mrepr : miustr → string
+| [] := ""
+| (c::cs) := c.repr ++ (miustr.mrepr cs)
+
+instance miurepr : has_repr miustr :=
+⟨λ u, u.mrepr⟩ 
+
+/--
+In the other direction, we set up coercion from `string` to `miustr`.
+-/
+def lchar_to_miustr : (list char) → miustr 
+| [] := []
+| (c::cs) :=
+  let ms := lchar_to_miustr cs in
+  match c with
+  | 'M' := M::ms
+  | 'I' := I::ms
+  | 'U' := U::ms
+  |  _  := []
+  end
+
+instance string_coe_miustr : has_coe string miustr :=
+⟨λ st, lchar_to_miustr st.data ⟩
+
+
+
+/-!
+### The rules of inference
+
+There are four rules of inference for MIU.
+
+Rule 1:  xI → xIU
+Rule 2:  Mx → Mxx
+Rule 3:  xIIIy → xUy
+Rule 4:  xUUy → xy
+
+For pedagogical purposes, we give definitons for the rules independently of the notion of
+derivability. We do not need these defintions to prove our main results. 
```

Comment:
```suggestion
derivability. We do not need these definitions to prove our main results. 
```

Reply:
I unconvinced that setting up the rules twice over makes for better pedagogy. Why not just do these examples after defining `derivable`?

Reply:
OK. I've changed this as suggested. Good idea!